### PR TITLE
Collection deserialisation needs to check for null tokens

### LINKED ISSUE #3813: Queries with "null" in JSON results cannot be deserialized
**NEST/Elasticsearch.Net version**: 7.0.0-alpha2

**Elasticsearch version**: 7.0.1

**Description of the problem including expected versus actual behavior**:
I am using `client.QuerySql()` to fetch data (configured externally). When the resulting data contains a `null` token, the deserialization fails with the following stack trace.
```
Elasticsearch.Net.UnexpectedElasticsearchClientException: expected:',', actual:'null', at offset:4450 ---> Elasticsearch.Net.JsonParsingException: expected:',', actual:'null', at offset:4450
   at Elasticsearch.Net.JsonReader.ReadIsValueSeparatorWithVerify()
   at Elasticsearch.Net.JsonReader.ReadIsEndArrayWithSkipValueSeparator(Int32& count)
   at Elasticsearch.Net.CollectionFormatterBase`4.Deserialize(JsonReader& reader, IJsonFormatterResolver formatterResolver)
   at Nest.SqlRowFormatter.Deserialize(JsonReader& reader, IJsonFormatterResolver formatterResolver)
   at Elasticsearch.Net.CollectionFormatterBase`4.Deserialize(JsonReader& reader, IJsonFormatterResolver formatterResolver)
   at Deserialize(Object[] , JsonReader& , IJsonFormatterResolver )
   at Elasticsearch.Net.JsonSerializer.Deserialize[T](Byte[] bytes, Int32 offset, IJsonFormatterResolver resolver)
   at Elasticsearch.Net.JsonSerializer.Deserialize[T](Stream stream, IJsonFormatterResolver resolver)
   at Nest.InternalSerializer.Deserialize[T](Stream stream)
   at Elasticsearch.Net.ResponseBuilder.SetBody[TResponse](ApiCallDetails details, RequestData requestData, Stream responseStream, String mimeType)
   at Elasticsearch.Net.ResponseBuilder.ToResponse[TResponse](RequestData requestData, Exception ex, Nullable`1 statusCode, IEnumerable`1 warnings, Stream responseStream, String mimeType)
   at Elasticsearch.Net.HttpConnection.Request[TResponse](RequestData requestData)
   at Elasticsearch.Net.RequestPipeline.CallElasticsearch[TResponse](RequestData requestData)
   at Elasticsearch.Net.Transport`1.Request[TResponse](HttpMethod method, String path, PostData data, IRequestParameters requestParameters)
   --- End of inner exception stack trace ---
   at Elasticsearch.Net.Transport`1.Request[TResponse](HttpMethod method, String path, PostData data, IRequestParameters requestParameters)
   at Nest.ElasticClient.QuerySql(Func`2 selector)
   at ElasticActor.Core.Worker.ExecuteTask(ActivatedTask task)
```

Am I missing a config setting or is this an oversight in development?

**Steps to reproduce**:
 1. Query data using SQL with `ElasticClient.QuerySql(...)`
 2. Exception gets thrown

### PR DESCRIPTION
**Pull request by codebrain**
Sql query results are composed of an array of values, with elements that may be null.

Fixes #3813

### CODE REVIEW

**Comment by russcam** [comment]
Spoke with @codebrain about this change. Will update this PR with a different approach

**Comment by codebrain** [comment]
Integration test currently failing: `Machine Learning - PreviewDatafeedApi`

File: src/Nest/Search/MultiSearch/MultiSearchResponseBuilder.cs [hunk 1]
```diff
@@ -13,8 +13,9 @@ internal class MultiSearchResponseBuilder : CustomResponseBuilderBase
 
 		public override object DeserializeResponse(IElasticsearchSerializer builtInSerializer, IApiCallDetails response, Stream stream)
 		{
-			var statefulSerializer = builtInSerializer.CreateStateful(Formatter);
-			return statefulSerializer.Deserialize<MultiSearchResponse>(stream);
+			return response.Success
```
Comment (russcam) [inline src/Nest/Search/MultiSearch/MultiSearchResponseBuilder.cs]: I don't think this should ever return null

File: src/Nest/Search/MultiSearch/MultiSearchResponseBuilder.cs [hunk 2]
```diff
@@ -24,8 +25,9 @@ public override object DeserializeResponse(IElasticsearchSerializer builtInSeria
 			CancellationToken ctx = default
 		)
 		{
-			var statefulSerializer = builtInSerializer.CreateStateful(Formatter);
-			return await statefulSerializer.DeserializeAsync<MultiSearchResponse>(stream, ctx);
+			return response.Success
```
Comment (russcam) [inline src/Nest/Search/MultiSearch/MultiSearchResponseBuilder.cs]: I don't think this should ever return null

File: src/Nest/XPack/MachineLearning/PreviewDatafeed/PreviewDatafeedResponseBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace Nest
+{
+	internal class PreviewDatafeedResponseBuilder<TDocument> : CustomResponseBuilderBase where TDocument : class
+	{
+		public static PreviewDatafeedResponseBuilder<TDocument> Instance { get; } = new PreviewDatafeedResponseBuilder<TDocument>();
+
+		public override object DeserializeResponse(IElasticsearchSerializer builtInSerializer, IApiCallDetails response, Stream stream)
+		{
+			return response.Success
```
Comment (russcam) [inline src/Nest/XPack/MachineLearning/PreviewDatafeed/PreviewDatafeedResponseBuilder.cs]: I don't think this should ever return null

File: src/Nest/XPack/MachineLearning/PreviewDatafeed/PreviewDatafeedResponseBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+		}
+
+		public override async Task<object> DeserializeResponseAsync(
+			IElasticsearchSerializer builtInSerializer,
+			IApiCallDetails response,
+			Stream stream,
+			CancellationToken ctx = default
+		)
+		{
+			return response.Success
```
Comment (russcam) [inline src/Nest/XPack/MachineLearning/PreviewDatafeed/PreviewDatafeedResponseBuilder.cs]: I don't think this should ever return null

File: src/Nest/XPack/Sql/TranslateSql/TranslateSqlResponseBuilder.cs [hunk 5]
```diff
@@ -11,8 +11,9 @@ internal class TranslateSqlResponseBuilder : CustomResponseBuilderBase
 
 		public override object DeserializeResponse(IElasticsearchSerializer builtInSerializer, IApiCallDetails response, Stream stream)
 		{
-			var result = builtInSerializer.Deserialize<ISearchRequest>(stream);
-			return new TranslateSqlResponse { Result = result };
+			return response.Success
```
Comment (russcam) [inline src/Nest/XPack/Sql/TranslateSql/TranslateSqlResponseBuilder.cs]: I don't think this should ever return null

File: src/Nest/XPack/Ssl/GetCertificates/GetCertificatesResponseBuilder.cs [hunk 6]
```diff
@@ -11,8 +11,9 @@ internal class GetCertificatesResponseBuilder : CustomResponseBuilderBase
 
 		public override object DeserializeResponse(IElasticsearchSerializer builtInSerializer, IApiCallDetails response, Stream stream)
 		{
-			var result = builtInSerializer.Deserialize<ClusterCertificateInformation[]>(stream);
-			return new GetCertificatesResponse { Certificates = result };
+			return response.Success
```
Comment (russcam) [inline src/Nest/XPack/Ssl/GetCertificates/GetCertificatesResponseBuilder.cs]: I don't think this should ever return null

File: src/Nest/XPack/Ssl/GetCertificates/GetCertificatesResponseBuilder.cs [hunk 7]
```diff
@@ -22,8 +23,9 @@ public override object DeserializeResponse(IElasticsearchSerializer builtInSeria
 			CancellationToken ctx = default
 		)
 		{
-			var result = await builtInSerializer.DeserializeAsync<ClusterCertificateInformation[]>(stream);
-			return new GetCertificatesResponse { Certificates = result };
+			return response.Success
```
Comment (russcam) [inline src/Nest/XPack/Ssl/GetCertificates/GetCertificatesResponseBuilder.cs]: I don't think this should ever return null

File: src/Nest/Cluster/NodesHotThreads/NodeHotThreadsResponseBuilder.cs [hunk 8]
```diff
@@ -49,8 +49,11 @@ where matches.Success
 			return new NodesHotThreadsResponse(info.ToList());
 		}
 
-		public override object DeserializeResponse(IElasticsearchSerializer builtInSerializer, IApiCallDetails response, Stream stream) 
+		public override object DeserializeResponse(IElasticsearchSerializer builtInSerializer, IApiCallDetails response, Stream stream)
 		{
+			if (response.Success == false)
```
Comment (russcam) [inline src/Nest/Cluster/NodesHotThreads/NodeHotThreadsResponseBuilder.cs]:
```suggestion
			if (!response.Success)
```

**Comment by codebrain** [comment]
Check for deserialisation of response in the event of an error

File: src/Elasticsearch.Net/Utf8Json/Formatters/CollectionFormatters.cs [hunk 9]
```diff
@@ -245,7 +245,8 @@ public TCollection Deserialize(ref JsonReader reader, IJsonFormatterResolver for
 				reader.ReadIsBeginArrayWithVerify();
 				while (!reader.ReadIsEndArrayWithSkipValueSeparator(ref count))
 				{
-					Add(ref builder, count - 1, formatter.Deserialize(ref reader, formatterResolver));
+					var value = formatter.Deserialize(ref reader, formatterResolver);
+					Add(ref builder, count - 1, value);
```
[the remark this answers is not in the document]
Comment (codebrain) [reply src/Elasticsearch.Net/Utf8Json/Formatters/CollectionFormatters.cs]: ++

File: src/Nest/CommonAbstractions/LazyDocument/LazyDocumentFormatter.cs [hunk 10]
```diff
@@ -25,7 +25,10 @@ public void Serialize(ref JsonWriter writer, ILazyDocument value, IJsonFormatter
 		public ILazyDocument Deserialize(ref JsonReader reader, IJsonFormatterResolver formatterResolver)
 		{
 			if (reader.GetCurrentJsonToken() == JsonToken.Null)
+			{
+				//reader.ReadNext();
```
[the remark this answers is not in the document]
Comment (codebrain) [reply src/Nest/CommonAbstractions/LazyDocument/LazyDocumentFormatter.cs]: Was part of a WIP, so have removed.

File: src/Nest/CommonAbstractions/LazyDocument/LazyDocumentFormatter.cs [hunk 11]
```diff
@@ -109,7 +112,10 @@ public void Serialize(ref JsonWriter writer, LazyDocument value, IJsonFormatterR
 		public LazyDocument Deserialize(ref JsonReader reader, IJsonFormatterResolver formatterResolver)
 		{
 			if (reader.GetCurrentJsonToken() == JsonToken.Null)
+			{
+				//reader.ReadNext();
```
[the remark this answers is not in the document]
Comment (codebrain) [reply src/Nest/CommonAbstractions/LazyDocument/LazyDocumentFormatter.cs]: Was part of a WIP, so have removed.

**Pull request merged**